pennylane/qnodes/base.py

```diff
@@ -131,6 +131,96 @@ def decompose_queue(ops, device):
     return new_ops
 
 
+def _compare_objects(object1, object2):
+    """Checks if two objects are equal.
+
+    Supports equality checks for NumPy arrays and TensorFlow/PyTorch tensors without requiring an
+    import of PyTorch or TensorFlow.
+
+    The function uses a broad try/except, which returns False on the exception. This allows us to
+    say that two incompatible objects are not equal without throwing an exception.
+
+    Args:
+        object1: first item to be compared
+        object2: second item to be compared
+
+    Returns:
+        bool: whether the two items are equal
+    """
+    try:
+        if type(object1) != type(object2):
+            return False
+
+        comparison = object1 == object2
+
+        if isinstance(comparison, bool):
+            return comparison
+
+        try:
+            return comparison.all()
+        except AttributeError:
+            return comparison.numpy().all()
+    except Exception:
```

Comment:
What could be an exception here?

Reply:
:thinking: here I'm paranoid, since I don't want the `evaluate()` method to break needlessly - we can just return false and re-evaluate the QNode if we're not sure.

But more concretely, what could go wrong here? Maybe
`object1 == object2`
or potentially
`comparison.all()` raising an exception other than `AttributeError`... or `comparison.numpy().all()` (note that this is the TF case) not working.